File: data/src/main/java/org/cryptomator/data/cloud/local/storageaccessframework/RootLocalStorageAccessFolder.java
package org.cryptomator.data.cloud.local.storageaccessframework;

import android.os.Build;
import android.provider.DocumentsContract;

import org.cryptomator.domain.Cloud;
import org.cryptomator.domain.LocalStorageCloud;

import androidx.annotation.RequiresApi;

import static android.net.Uri.parse;

@RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
public class RootLocalStorageAccessFolder extends LocalStorageAccessFolder {

	private final LocalStorageCloud localStorageCloud;

	public RootLocalStorageAccessFolder(LocalStorageCloud localStorageCloud) {
		super(null, //
				"", //
				"", //
				DocumentsContract.getTreeDocumentId( //
						parse(localStorageCloud.rootUri())), //
				DocumentsContract.buildChildDocumentsUriUsingTree( //
						parse(localStorageCloud.rootUri()), //
						DocumentsContract.getTreeDocumentId( //
								parse(localStorageCloud.rootUri())))
						.toString());
		this.localStorageCloud = localStorageCloud;
	}

	@Override
	public Cloud getCloud() {
		return localStorageCloud;
	}

	@Override
	public LocalStorageAccessFolder withCloud(Cloud cloud) {
		return new RootLocalStorageAccessFolder((LocalStorageCloud) cloud);
	}
}

File: data/src/main/java/org/cryptomator/data/cloud/local/RootLocalStorageAccessFolder.kt
package org.cryptomator.data.cloud.local

import android.net.Uri
import android.provider.DocumentsContract
import org.cryptomator.domain.Cloud
import org.cryptomator.domain.LocalStorageCloud

class RootLocalStorageAccessFolder(private val localStorageCloud: LocalStorageCloud) : LocalStorageAccessFolder(
	null,  //
	"",  //
	"",  //
	DocumentsContract.getTreeDocumentId(Uri.parse(localStorageCloud.rootUri())),  //
	DocumentsContract.buildChildDocumentsUriUsingTree( //
		Uri.parse(localStorageCloud.rootUri()),  //
		DocumentsContract.getTreeDocumentId(Uri.parse(localStorageCloud.rootUri()))
	).toString()
) {

	override val cloud: Cloud
		get() = localStorageCloud

	override fun withCloud(cloud: Cloud?): RootLocalStorageAccessFolder {
		return RootLocalStorageAccessFolder(cloud as LocalStorageCloud)
	}
}
